Migrating to use native Pytorch AMP

sjrl opened the pull request:
**Related Issue(s)**:  [Issue #1512 ](https://github.com/deepset-ai/haystack/issues/1512) [Issue #1222](https://github.com/deepset-ai/haystack/issues/1222)

**Proposed changes**:
Migrating to Pytorch's native AMP https://pytorch.org/docs/stable/notes/amp_examples.html because it is much easier to use (no additional dependency on apex) and needs fewer code changes to support and it's recommended (https://discuss.pytorch.org/t/torch-cuda-amp-vs-nvidia-apex/74994). 

Using the native AMP support in Pytorch requires using [torch.cuda.amp.autocast](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.autocast) and [torch.cuda.amp.GradScaler](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler) together. These can easily be "turned off" so no autocasting or scaling occurs by passing the option `enabled=False`. 

For example, the following code performs a **standard** training loop because we have passed the option `enabled=False` to both `autocast` and `GradScaler`

```python
use_amp = False

# Creates model and optimizer in default precision
model = Net().cuda()
optimizer = optim.SGD(model.parameters(), ...)

# Creates a GradScaler once at the beginning of training.
scaler = GradScaler(enabled=use_amp)

for epoch in epochs:
    for input, target in data:
        optimizer.zero_grad()

        # Runs the forward pass with autocasting.
        with autocast(enable=use_amp):
            output = model(input)
            loss = loss_fn(output, target)

        # Scales loss.  Calls backward() on scaled loss to create scaled gradients.
        # Backward passes under autocast are not recommended.
        # Backward ops run in the same dtype autocast chose for corresponding forward ops.
        scaler.scale(loss).backward()

        # scaler.step() first unscales the gradients of the optimizer's assigned params.
        # If these gradients do not contain infs or NaNs, optimizer.step() is then called,
        # otherwise, optimizer.step() is skipped.
        scaler.step(optimizer)

        # Updates the scale for next iteration.
        scaler.update()
```
And similarly it is easy to turn on AMP by passing `enabled=True`.

There are **breaking changes**:
1. The input type for `use_amp` in the `FARMReader.train` method is now a type `bool` instead of type `str`. 
2. Currently this PR deprecates apex and does not try to support Pytorch AMP and Apex AMP.

## Pre-flight checklist
- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/master/CONTRIBUTING.md)
- [x] If this is a code change, I added tests or updated existing ones 
- [x] If this is a code change, I updated the docstrings


## TODO
- [x] Checked that `FARMReader.train` works with `use_amp=True` (and `False`) on a single GPU
- [x] Test tutorial 9 DPR training in GPU environment with `use_amp` turned on (and off) and with `grad_acc_steps`
- [x] Add `use_amp` to trainer state dict so when restarting from a checkpoint AMP is set up as expected
- [ ] Test multi-gpu training with AMP
  - [ ] Works with torch.nn.DistributedDataParallel (Link to [docs](https://pytorch.org/docs/stable/notes/amp_examples.html#distributeddataparallel-one-gpu-per-process))
  - [x] Works with torch.nn.DataParallel. Tested on 4 GPUs (g4dn.12xlarge instance), confirmed usage using `nvidia-smi`. It appears the "apply autocast as part of your model’s forward method to ensure it’s enabled in side threads." statement only refers to when we use multiple GPUs per process. Docs [here](https://pytorch.org/docs/stable/notes/amp_examples.html#distributeddataparallel-multiple-gpus-per-process). From what I understand I believe we only use one GPU per process which is [recommended](https://pytorch.org/docs/stable/notes/amp_examples.html#dataparallel-in-a-single-process).  
- [ ] Open PR in https://github.com/deepset-ai/haystack-website editing this [file](https://github.com/deepset-ai/haystack-website/blob/source/docs/latest/guides/optimization.mdx) explaining AMP
  -  I could imagine a section on that page where you briefly describe what AMP is, when to use it and how to use it. Perhaps also give readers an idea of how big a trade off in accuracy / speed it is.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/deepset-ai/haystack?pullRequest=2827) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/deepset-ai/haystack?pullRequest=2827) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/deepset-ai/haystack?pullRequest=2827) it.</sub>

sjrl commented on haystack/modeling/model/optimization.py:
```diff
@@ -300,7 +294,7 @@ def optimize_model(
     optimizer=None,
     distributed: Optional[bool] = False,
-    use_amp: Optional[str] = None,
+    use_amp: bool = False,
```
This parameter is no longer needed. Is it safe for me to remove it? Or do I need to add some deprecation warning?

sjrl replied:
This sounds good to me! I'll make the changes by adding the warning and the mapping.

sjrl commented on haystack/modeling/training/base.py:
```diff
@@ -372,30 +372,27 @@ def train(self):
+        return self.backward_propagate(loss, step)
 
     def backward_propagate(self, loss: torch.Tensor, step: int):
-        loss = self.adjust_loss(loss)
```
I needed to move this out of `backward_propagate` because Pytorch shows that the `adjust_loss` step should happen within the same `autocast` context. Link to explanation: https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-accumulation

julian-risch commented:
First of all, I support removing nvidia apex and adding pytorch amp. 👍 Doing quick research regarding this decision, it's what seems to be the preferred, future-proof way: https://discuss.pytorch.org/t/torch-cuda-amp-vs-nvidia-apex/74994/2
What we should once this PR is ready to be merged is a small benchmark. And we would also need to ensure that the documentation explains when to use this feature, how to use this feature and what to expect from it.

sjrl commented:
> And we would also need to ensure that the documentation explains when to use this feature, how to use this feature and what to expect from it.

Where should this documentation be added?

sjrl commented on haystack/modeling/model/optimization.py:
```diff
@@ -161,13 +133,13 @@ def initialize_optimizer(
         schedule_opts["num_training_steps"] = num_train_optimization_steps
 
     # Log params
-    tracker.track_params({"use_amp": use_amp, "num_train_optimization_steps": schedule_opts["num_training_steps"]})
```
I moved tracking of this parameter into the Trainer since `use_amp` is no longer needed or used by the optimizer. Is that alright?

sjrl commented on haystack/modeling/model/optimization.py:
```diff
@@ -119,20 +96,15 @@ def initialize_optimizer(
-                    "O3" (Pure FP16).
-                    See details on: https://nvidia.github.io/apex/amp.html
+    :param use_amp: Whether to use automatic mixed precision (AMP) natively implemented in PyTorch.
+                    Find more information at https://pytorch.org/docs/stable/amp.html
```
In addition to the `logger.warning` in the function does it make sense to update the docstring to explain that `use_amp` is no longer used by this function?

julian-risch replied:
Yes, please. 👍

julian-risch commented:
> > And we would also need to ensure that the documentation explains when to use this feature, how to use this feature and what to expect from it.
> 
> Where should this documentation be added?

@brandenchan Could you please advise here? Thank you! I would say here: https://haystack.deepset.ai/guides/optimization

sjrl commented:
@julian-risch Do we happen to have any GPU resources in our CI testing so I could add a test to make sure the changes to `use_amp` work as expected?

julian-risch commented:
> @julian-risch Do we happen to have any GPU resources in our CI testing so I could add a test to make sure the changes to `use_amp` work as expected?

Unfortunately the CI is CPU only, no. I would recommend an EC2 instance where you could checkout your branch and run the tests while using the GPU.

brandenchan commented:
> > > And we would also need to ensure that the documentation explains when to use this feature, how to use this feature and what to expect from it.
> > 
> > 
> > Where should this documentation be added?
> 
> @brandenchan Could you please advise here? Thank you! I would say here: https://haystack.deepset.ai/guides/optimization So that would require a separate PR in https://github.com/deepset-ai/haystack-website but maybe @brandenchan can help with that.

Hi yes sure! To make changes to that specific optimisation page, you will want to open a PR that edits [this file](https://github.com/deepset-ai/haystack-website/blob/source/docs/latest/guides/optimization.mdx). I could imagine a section on that page where you briefly describe what AMP is, when to use it and how to use it. Perhaps also give readers an idea of how big a trade off in accuracy / speed it is. 

Happy to review it once you've opened the PR! Just let me know if you want any more pointers regarding this process. The readme of the haystack-website repo describes how you can set up a preview environment on your own machine. There is also an action in the PR which will deploy your changes to a staging environment though this takes a bit longer than previewing on your own machine.

julian-risch commented:
Hi @sjrl have you already created a PR regarding documentation? This PR here looks good to me on first sight. It's a bit unfortunate that there are so many unrelated doc string updates or whitespace changes in there but it's not a big problem. Before I go through the code in detail, I would like to try it out a bit and in that context a draft PR regarding documentation would be great. Did you test it out yourself already?

sjrl commented:
Hi @julian-risch, sorry about the additional doc updates. I was just making those as I was exploring the code. I'll put them into a separate PR next time. 

I have had a chance to test this code out for training a FARMReader on a single GPU. I haven't tried the different combinations (e.g. multi-gpu, gradient_clipping, etc.) to make sure all features work as expected.

Also I have not had time yet to make a draft PR for documentation.

julian-risch commented:
Alright, thanks for the quick response. No worries about the doc string updates. It's better to have them here than not having them at all! 😄

julian-risch commented on docs/_src/api/api/reader.md:
```diff
@@ -520,7 +500,7 @@ Example:
 
 - `question`: Question string
-- `documents`: List of documents as string type
+- `texts`: List of documents as string type
```
As documents have their own type, I would prefer something like "list of texts as string type" here.

julian-risch commented on haystack/modeling/model/optimization.py:
```diff
@@ -119,20 +95,17 @@ def initialize_optimizer(
+        logger.warning(
+            "Only PyTorch automatic mixed precision is supported. The Apex library is no longer supported.\n"
+            "This means that use_amp is no longer used by modeling.model.initialize_optimizer since it is not needed\n"
+            "to initialize native PyTorch automatic mixed precision."
```
A link to the documentation page about optimization should be added here, so that interested users can get more info. I'd say most important for them to know is that automatic mixed precision works without further ado if they pass `use_amp=True`.
I would suggest to only show the warning message if `use_amp` is a string. If it is a boolean, the user has changed the use of the parameter and already knows about the change.

julian-risch commented on haystack/modeling/model/optimization.py:
```diff
@@ -310,24 +280,23 @@ def optimize_model(
     """
-    model, optimizer = _init_amp(model, device, optimizer, use_amp)
+    if use_amp:
+        logger.warning(
```
I have the same comment here as for the warning above. 🙂

julian-risch replied:
Do you think we need the warning both in `initialize_optimizer` and `optimize_model`?

sjrl replied:
I think it's nice to have in the off chance that a user decides to call `initialize_optimizer` and `optimize_model` explicitly instead of just using the built-in `train` methods.

Without the warning message, I would recommend we remove `use_amp` here and list this as a breaking change.

julian-risch replied:
Alright, I think it's okay to show the warning twice.

julian-risch commented on haystack/modeling/model/predictions.py:
```diff
@@ -294,7 +294,7 @@ def _answers_to_json(self, ext_id, squad=False) -> List[Dict]:
         Convert all answers into a json format
 
-        :param id: ID of the question document pair
+        :param ext_id: ID of the question document pair
```
Good catch 👍

julian-risch commented on haystack/modeling/training/base.py:
```diff
@@ -156,8 +149,8 @@ def __init__(
-        :param use_amp: Whether to use automatic mixed precision with Apex. One of the optimization levels must be chosen.
-                        "O1" is recommended in almost all cases.
+        :param use_amp: Whether to use automatic mixed precision (AMP) natively implemented in PyTorch.
+                        Find more information at https://pytorch.org/docs/stable/amp.html
```
let's add the link to our own documentation, in particular the optimization page, as well.

julian-risch commented on haystack/modeling/training/base.py:
```diff
@@ -180,16 +173,30 @@ def __init__(
         self.model = model
         self.data_silo = data_silo
         self.epochs = int(epochs)
+        if isinstance(use_amp, str):
```
This is the if statement I would also add to the other warnings above.

julian-risch commented on haystack/modeling/training/base.py:
```diff
@@ -180,16 +173,30 @@ def __init__(
+                    f"Since you provided Apex optimization level {use_amp}, automatic mixed precision has been set to {amp_mapping[use_amp]}.\n"
+                    "In the future provide use_amp=True to turn on automatic mixed precision."
+                )
+                use_amp = amp_mapping[use_amp]
```
Great fallback solution 👍

julian-risch commented on haystack/nodes/reader/farm.py:
```diff
@@ -1248,7 +1226,7 @@ def predict_on_texts(self, question: str, texts: List[str], top_k: Optional[int]
 
         :param question: Question string
-        :param documents: List of documents as string type
+        :param texts: List of documents as string type
```
As above, I wouldn't call it "documents" here as they have their own type. Let's call it texts instead or maybe document texts.

julian-risch commented on haystack/nodes/retriever/dense.py:
```diff
@@ -607,18 +607,17 @@ def train(
         :param save_dir: directory where models are saved
         :param query_encoder_save_dir: directory inside save_dir where query_encoder model files are saved
         :param passage_encoder_save_dir: directory inside save_dir where passage_encoder model files are saved
+        :param checkpoint_root_dir:
```
Let's add brief descriptions of these three parameters here.

julian-risch commented on haystack/nodes/retriever/dense.py:
```diff
@@ -1272,19 +1270,18 @@ def train(
         :param query_encoder_save_dir: Directory inside save_dir where query_encoder model files are saved.
         :param passage_encoder_save_dir: Directory inside save_dir where passage_encoder model files are saved.
         :param table_encoder_save_dir: Directory inside save_dir where table_encoder model files are saved.
+        :param checkpoint_root_dir:
```
Let's add brief descriptions of these three parameters here.

julian-risch commented on haystack/pipelines/base.py:
```diff
@@ -1072,7 +1072,7 @@ def eval(
 
         :param labels: The labels to evaluate on
-        :param documents: List of List of Document that the first node in the pipeline should get as input per multilabel. Can be used to evaluate a pipeline that consists of a reader without a retriever.
+        :param documents: List of Document that the first node in the pipeline should get as input per multilabel. Can be used to evaluate a pipeline that consists of a reader without a retriever.
```
Might be confusing whether the parameter is now a list only or a list of lists. I'd suggest to change it to: "List of List of Document, where the input to the first node in the pipeline is one List of Document per MultiLabel"

sjrl replied:
RIght, I think it should only be a List of List of Document objects.

julian-risch reviewed (changes requested):
This PR is already looking great! 👍 The fallback solution to map the old apex string parameters to boolean is nice for the users. 
I tested the speed up and output on a single GPU myself. Works for me. 👍 What I didn't test is multiple GPUs. 
I would like to request a few smaller changes in docstrings and in the code I think we only need to show the warning when the provided parameter is not of type boolean but string. The rest of the code looks good to me. Good job!

sjrl commented on docs/_src/api/api/pipelines.md:
```diff
@@ -526,7 +526,7 @@ Some of these scopes require additional information that already needs to be spe
 
 - `labels`: The labels to evaluate on
-- `documents`: List of List of Document that the first node in the pipeline should get as input per multilabel. Can be used to evaluate a pipeline that consists of a reader without a retriever.
+- `documents`: List of Document that the first node in the pipeline should get as input per multilabel. Can be used to evaluate a pipeline that consists of a reader without a retriever.
```
Thanks for pointing this out! I should have looked at the type annotation. Looking at the code for `eval` it appears only a list of list of documents is allowed so I'll change it back, but without the second capitalization of list.

sjrl commented:
I just found some additional docs on using AMP with multiple GPUs. 
* For `DistributedDataParallel` it looks like the existing code should work (Link to [docs](https://pytorch.org/docs/stable/notes/amp_examples.html#distributeddataparallel-one-gpu-per-process)
* For `torch.nn.DataParallel` it looks like we would need to "apply autocast as part of your model’s forward method to ensure it’s enabled in side threads." (Link to [docs]([torch.nn.DataParallel](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html#torch.nn.DataParallel)))

sjrl commented on docs/_src/api/api/reader.md:
```diff
@@ -273,8 +261,7 @@ that gets split off from training data for eval.
 Unused if `use_gpu` is False.
-- `student_batch_size`: Number of samples the student model receives in one batch for training
-- `student_batch_size`: Number of samples the teacher model receives in one batch for distillation
+- `batch_size`: Number of samples the student model and teacher model receives in one batch for training
```
I believe it should be `receives` to match the plurality of `Number of samples`

agnieszka-m commented:
Sorry for the late review, just realized I forgot to click "finish your review" after I reviewed it..

sjrl commented on docs/_src/api/api/pipelines.md:
```diff
@@ -526,7 +526,7 @@ Some of these scopes require additional information that already needs to be spe
 
 - `labels`: The labels to evaluate on
-- `documents`: List of List of Document that the first node in the pipeline should get as input per multilabel. Can be used to evaluate a pipeline that consists of a reader without a retriever.
+- `documents`: List of Document that the first node in the pipeline should get as input per multilabel. Can be used to evaluate a pipeline that consists of a reader without a retriever.
```
I've incorporated your changes and Julian's suggested changes.

sjrl commented on docs/_src/api/api/reader.md:
```diff
@@ -141,14 +141,8 @@ Options for different schedules are available in FARM.
-"O3" (Pure FP16).
-See details on: https://nvidia.github.io/apex/amp.html
+- `use_amp`: Whether to use automatic mixed precision (AMP) natively implemented in PyTorch.
+Find more information at https://pytorch.org/docs/stable/amp.html
```
@agnieszka-m Do you know how to hyperlink text docstrings? Is it the same format as GitHub using `[text](url)`?

agnieszka-m replied:
I don't know to be honest..

sjrl replied:
After some digging it does look like our docs page does support hyperlinks in the same way as Github. We can see some examples on this doc page: https://haystack.deepset.ai/reference/query-classifier

sjrl commented on haystack/modeling/model/optimization.py:
```diff
@@ -219,25 +192,22 @@ def _get_optim(model, opts: Dict):
-                        f"Optimizer '{optimizer_name}' not found in 'torch', 'transformers', 'apex' or 'local imports"
-                    )
+                raise AttributeError(
+                    f"Optimizer '{optimizer_name}' not found in 'torch', 'transformers' or 'local imports"
```
Unfortunately, not that I can easily think of.

julian-risch commented:
Hi @sjrl what's the status of this PR? Would you like to get another round of reviews from @agnieszka-m and me?

sjrl commented:
Hi, @julian-risch thanks for checking in. I don't think there are any new changes since the last review other than the merging of main. I hope to be able to finish the task of checking this on multiple GPUs next week.

julian-risch commented:
Hi @sjrl we are planning a release in the next two weeks. Could this PR maybe make it into the new release? Did you have the chance to test multi-gpu training with AMP? In the todo list there is another open item "For torch.nn.DataParallel it looks like we would need to "apply autocast as part of your model’s forward method to ensure it’s enabled in side threads.""

julian-risch commented:
FYI: we might upgrade to torch 1.13.1 once it's released.

sjrl commented:
> Hi @sjrl we are planning a release in the next two weeks. Could this PR maybe make it into the new release? Did you have the chance to test multi-gpu training with AMP? In the todo list there is another open item "For torch.nn.DataParallel it looks like we would need to "apply autocast as part of your model’s forward method to ensure it’s enabled in side threads.""

Hey @julian-risch. I haven't had time to test the multi-gpu training with `torch.nn.DataParallel`. I have tested `torch.nn.DistributedDataParallel` which does work. I can try to get this done before the next release, but I am fairly busy at the moment so I'd also be happy to receive help here if you have the time.

sjrl commented:
Hey @julian-risch I first tried to get `DistributedDataParallel` to work today by turning on `distributed` training (without amp). Right now this option is hardcoded to be off since we are using the default value in the call to `initialize_optimizer`
https://github.com/deepset-ai/haystack/blob/77cea8b1408699f69dc09420f9f9af3cf5e6f47a/haystack/modeling/model/optimization.py#L71-L79

And even trying to set this to true when using `initialize_optimizer` in the training loop https://github.com/deepset-ai/haystack/blob/77cea8b1408699f69dc09420f9f9af3cf5e6f47a/haystack/nodes/reader/farm.py#L266

causes a multiprocessing error. So it looks like we would need to first fix the distributed training feature before confirming that amp works with it as well.

julian-risch commented:
So what's your opinion on the best way forward? I'd say we merge the changes that we have up to now and support just `torch.nn.DataParallel` but not `torch.nn.DistributedDataParallel`. Investigating the multiprocessing error and supporting `torch.nn.DistributedDataParallel` should then become the topic of a separate issue that we can add to the backlog.

sjrl commented:
@julian-risch Yes, I completely agree.

julian-risch reviewed (approved):
Looks very good to me! 👍 Thanks for putting in the extra effort and adding a fast test!
